### PR TITLE
Retain ssl/tls config from seed uris in Master/Slave context

### PR DESCRIPTION
**Pull request by acmcelwee**
Still WIP, but I'm opening this to start a discussion.

The problem I'm trying to solve is the combination of an AWS Elasticache redis cluster (in master-slave cluster-mode) w/ their recently added in-transit encryption. We aren't currently using redis cluster because of the need for at least 3 primary nodes + any optional layer of replicas, and we're rather satisfied w/ the simplicity of master/slave replication and the minimal node count per cluster that it gives.

I'm experimenting to see if I can somehow get a static Master/Slave topology + SSL to work. Before this PR, it simply isn't possible to create a static master/slave topology w/ something like: `MasterSlave.connect(client, codec, Arrays.asList("rediss://foo0.bar.baz", "rediss://foo1.bar.baz"))`. The initial connections all fail because the ssl/tls-related uri info is dropped in the conversion of the `RedisMasterSlaveNode` objects. 

Once I added that, the initial SSL handshake succeeded, but the connection didn't actually succeed, and I saw an endless stream of logs like this:
```
...
INFO  [2018-01-24T05:00:08.495Z] [] [] [lettuce-nioEventLoop-4-2]: io.lettuce.core.protocol.ReconnectionHandler: Reconnected to abc.xyz.nzcsol.use1.cache.amazonaws.com:6379
INFO  [2018-01-24T05:00:08.496Z] [] [] [lettuce-nioEventLoop-4-1]: io.lettuce.core.protocol.ReconnectionHandler: Reconnected to abc.xyz.nzcsol.use1.cache.amazonaws.com:6379
INFO  [2018-01-24T05:00:08.592Z] [] [] [lettuce-nioEventLoop-4-3]: io.lettuce.core.protocol.ReconnectionHandler: Reconnected to abc.xyz.nzcsol.use1.cache.amazonaws.com:6379
INFO  [2018-01-24T05:00:08.594Z] [] [] [lettuce-nioEventLoop-4-4]: io.lettuce.core.protocol.ReconnectionHandler: Reconnected to abc.xyz.nzcsol.use1.cache.amazonaws.com:6379
...
```

Digging in further, I found that the `MasterSlaveConnectionProvider` was ignoring any ssl/tls info present in the seed uri, and fixing _that_ finally let me create and connect to the static master/slave topology over ssl. Now, I see these log lines as the stateful connection objects are getting wired up at app startup time 
```
WARN  [2018-01-24T19:27:27.822Z] [] [] [lettuce-nioEventLoop-4-2]: io.lettuce.core.RedisChannelHandler: Connection is already closed
WARN  [2018-01-24T19:27:27.822Z] [] [] [lettuce-nioEventLoop-4-1]: io.lettuce.core.RedisChannelHandler: Connection is already closed
```

I create a few different stateful connections to the same cluster because each of the call sites have slightly different uses. 

## Open Questions
* ~~Does this feel like it's headed in the right direction?~~
* Any thoughts on the `Connection is already closed` warnings?
* ~~What branch should I work from for the final PR? We were currently on a 4.x release, but I went ahead and upgraded our dep to the `5.0.1-RELEASE` when I started investigating the failures to connect.~~
* ~~Test coverage - I obviously don't have any yet, but I want to get some input on the direction, first.~~
* Blind spots - what am I missing from this being my first time to dig into the actual library code? Do you see any immediate issues w/ this for dynamic master/slave topologies, sentinel topologies, etc?

### CODE REVIEW

**Comment by mp911de** [comment]
Awesome, thanks a lot for looking into that issue. Propagating SSL details was probably the only missing bit. 

I need to take a closer look on why `Connection is already closed` happens, in particular, why connections are closed more than one time. It should not be an issue per se, just something that happens too often.

Usually, pull requests go towards `master`, but either way is fine. `5.x` is the `master` for Lettuce 5 so I can pick it up and propagate to the other branches.

You can find SSL tests in `SslTest` that is guarded to not run if no SSL port is open. It would make sense to create a `MasterSlaveSslTest` for that particular case. All SSL and Redis instance setup happens through `Makefile`. It should be sufficient to add SSL tunnels for Redis running at `localhost:6482` and `localhost:6482`, these are a master and a slave. Although they are sentinel-monitored, they are fine to be used without Sentinel.

**Comment by acmcelwee** [comment]
Great. I'll start working on tests for this, and am happy to rebase the PR on master.

**Comment by acmcelwee** [comment]
@mp911de I've got a good start on the tests, and I just added to `SslTest` for now. It seemed a little more clear to keep all of the SSL-related test coverage together, but let me know if you want me to start breaking those out. Interestingly, I thought I could actually get away w/ removing the changes to `RedisMasterSlaveNode`, but those changes are necessary to allow the `pingBeforeActivateConnection` functionality.

I'll probably take a look at backporting this to the 4.x branch, as well.

The codecov bot tells me that I've increased the complexity, but it's hard to see how these changes truly increased the complexity.

**Comment by acmcelwee** [comment]
Also, it appears the "... Connection is already closed ..." loglines are most likely coming from a pub/sub context, and they only happen on the 5.x branch. 

```
WARN  [2018-01-24T19:27:27.822Z] [] [] [lettuce-nioEventLoop-4-2]: io.lettuce.core.RedisChannelHandler: Connection is already closed
WARN  [2018-01-24T19:27:27.822Z] [] [] [lettuce-nioEventLoop-4-1]: io.lettuce.core.RedisChannelHandler: Connection is already closed
```

File: Makefile [hunk 1]
```diff
@@ -269,6 +269,37 @@ work/stunnel.conf:
 	@echo cafile=$(ROOT_DIR)/work/ca/certs/ca.cert.pem >> $@
 	@echo verify=2 >> $@
 
+	@echo [stunnel-master-slave-node-1] >> $@
```
Comment (mp911de) [inline Makefile]: Having two tunnels should be sufficient

File: Makefile [hunk 2]
```diff
@@ -269,6 +269,37 @@ work/stunnel.conf:
...
+
+	@echo [stunnel-master-slave-node-2] >> $@
+	@echo accept = 127.0.0.1:8444 >> $@
+	@echo connect = 127.0.0.1:6483 >> $@
+	@echo cert=$(ROOT_DIR)/work/ca/certs/localhost.cert.pem >> $@
+	@echo key=$(ROOT_DIR)/work/ca/private/localhost.decrypted.key.pem >> $@
+	@echo capath=$(ROOT_DIR)/work/ca/certs/localhost.cert.pem >> $@
+	@echo cafile=$(ROOT_DIR)/work/ca/certs/localhost.cert.pem >> $@
+
+	@echo [stunnel-master-slave-client-cert-node-1] >> $@
```
Comment (mp911de) [inline Makefile]: Let's skip the cert-auth nodes. Certificate-based authentication isn't something we have to test again.

File: src/test/java/io/lettuce/core/cluster/RedisClusterPasswordSecuredSslTest.java [hunk 3]
```diff
@@ -47,7 +47,7 @@
     public static final String SLOT_1_KEY = "8HMdi";
     public static final String SLOT_16352_KEY = "UyAa4KqoWgPGKa";
 
-    public static RedisURI redisURI = RedisURI.builder().redis(host(), CLUSTER_PORT_SSL_1).withPassword("foobared")
+    public static RedisURI redisURI = RedisURI.Builder.redis(host(), CLUSTER_PORT_SSL_1).withPassword("foobared")
```
Comment (mp911de) [inline src/test/java/io/lettuce/core/cluster/RedisClusterPasswordSecuredSslTest.java]: Are these changes required?

File: src/test/java/io/lettuce/SslTest.java [hunk 4]
```diff
@@ -36,40 +40,70 @@
...
             .build();
 
-    private static final RedisURI URI_CLIENT_CERT_AUTH = RedisURI.Builder.redis(host(), sslPort(2)) //
-            .withSsl(true) //
+
+    private static final RedisURI URI_CLIENT_CERT_AUTH = sslURIBuilder(2) //
             .withVerifyPeer(true) //
             .build();
 
+    private static final List<RedisURI> MASTER_SLAVE_URIS_NO_VERIFY = Arrays.asList(
```
Comment (mp911de) [inline src/test/java/io/lettuce/SslTest.java]: Let's keep the tests simple by not re-testing functionality that is already covered, in particular, we should just have tests for the regular test cases with and without verifications. Client cert/OpenSSL/various trust-stores test cases are already covered and don't require own tests.

File: src/test/java/io/lettuce/SslTest.java [hunk 5]
```diff
@@ -217,6 +251,139 @@ public void sslWithVerificationWillFail() {
         redisClient.connect(redisUri).sync();
     }
 
+    @Test
+    public void masterSlaveWithSsl() {
```
Comment (mp911de) [inline src/test/java/io/lettuce/SslTest.java]: 👍

File: src/test/java/io/lettuce/SslTest.java [hunk 6]
```diff
@@ -217,6 +251,139 @@ public void sslWithVerificationWillFail() {
...
+
+        RedisCommands<String, String> connection =
+                MasterSlave.connect(redisClient, StringCodec.UTF8, MASTER_SLAVE_URIS_NO_VERIFY).sync();
+        connection.set("key", "value");
+        assertThat(connection.get("key")).isEqualTo("value");
+        connection.getStatefulConnection().close();
+    }
+
+    @Test
+    public void masterSlaveWithJdkSsl() {
```
Comment (mp911de) [inline src/test/java/io/lettuce/SslTest.java]: This one could be removed (see comment at top).

File: src/test/java/io/lettuce/SslTest.java [hunk 7]
```diff
@@ -217,6 +251,139 @@ public void sslWithVerificationWillFail() {
...
+                .build();
+        setOptions(sslOptions);
+
+        verifyMasterSlaveConnection(MASTER_SLAVE_URIS_VERIFY);
+    }
+
+    @Test
+    public void masterSlaveWithJdkSslUsingTruststoreUrl() throws Exception {
+
+        SslOptions sslOptions = SslOptions.builder() //
```
Comment (mp911de) [inline src/test/java/io/lettuce/SslTest.java]: 👍

File: src/test/java/io/lettuce/SslTest.java [hunk 8]
```diff
@@ -217,6 +251,139 @@ public void sslWithVerificationWillFail() {
...
+                .build();
+        setOptions(sslOptions);
+
+        verifyMasterSlaveConnection(MASTER_SLAVE_URIS_VERIFY);
+    }
+
+    @Test
+    public void masterSlaveWithClientCertificates() throws Exception {
+
+        SslOptions sslOptions = SslOptions.builder() //
```
Comment (mp911de) [inline src/test/java/io/lettuce/SslTest.java]: This one could be removed (see comment at top).

File: src/test/java/io/lettuce/SslTest.java [hunk 9]
```diff
@@ -217,6 +251,139 @@ public void sslWithVerificationWillFail() {
...
+        setOptions(sslOptions);
+
+        verifyMasterSlaveConnection(MASTER_SLAVE_URIS_CLIENT_CERT_AUTH);
+    }
+
+
+    @Test(expected = RedisConnectionException.class)
+    public void masterSlaveWithClientCertificatesWithoutKeystore() throws Exception {
+
+        SslOptions sslOptions = SslOptions.builder() //
```
Comment (mp911de) [inline src/test/java/io/lettuce/SslTest.java]: This one could be removed (see comment at top).

File: src/test/java/io/lettuce/SslTest.java [hunk 10]
```diff
@@ -217,6 +251,139 @@ public void sslWithVerificationWillFail() {
...
+                .jdkSslProvider() //
+                .truststore(TRUSTSTORE_FILE) //
+                .build();
+        setOptions(sslOptions);
+
+        verifyMasterSlaveConnection(MASTER_SLAVE_URIS_CLIENT_CERT_AUTH);
+    }
+
+    @Test(expected = RedisConnectionException.class)
+    public void masterSlaveWithJdkSslUsingTruststoreUrlWithWrongPassword() throws Exception {
```
Comment (mp911de) [inline src/test/java/io/lettuce/SslTest.java]: This one could be removed (see comment at top).

File: src/test/java/io/lettuce/SslTest.java [hunk 11]
```diff
@@ -217,6 +251,139 @@ public void sslWithVerificationWillFail() {
...
+                .jdkSslProvider() //
+                .truststore(truststoreURL(), "knödel") //
+                .build();
+        setOptions(sslOptions);
+
+        verifyMasterSlaveConnection(MASTER_SLAVE_URIS_VERIFY);
+    }
+
+    @Test(expected = RedisConnectionException.class)
+    public void masterSlaveWithJdkSslFailsWithWrongTruststore() {
```
Comment (mp911de) [inline src/test/java/io/lettuce/SslTest.java]: 👍

File: src/test/java/io/lettuce/SslTest.java [hunk 12]
```diff
@@ -217,6 +251,139 @@ public void sslWithVerificationWillFail() {
...
+        SslOptions sslOptions = SslOptions.builder() //
+                .jdkSslProvider() //
+                .build();
+        setOptions(sslOptions);
+
+        verifyMasterSlaveConnection(MASTER_SLAVE_URIS_VERIFY);
+    }
+
+    @Test
+    public void masterSlaveWithOpenSsl() {
```
Comment (mp911de) [inline src/test/java/io/lettuce/SslTest.java]: This one could be removed (see comment at top).

File: src/test/java/io/lettuce/SslTest.java [hunk 13]
```diff
@@ -217,6 +251,139 @@ public void sslWithVerificationWillFail() {
...
+                .openSslProvider() //
+                .truststore(TRUSTSTORE_FILE) //
+                .build();
+        setOptions(sslOptions);
+
+        verifyMasterSlaveConnection(MASTER_SLAVE_URIS_VERIFY);
+    }
+
+    @Test(expected = RedisConnectionException.class)
+    public void masterSlaveWithOpenSslFailsWithWrongTruststore() {
```
Comment (mp911de) [inline src/test/java/io/lettuce/SslTest.java]: This one could be removed (see comment at top).

File: src/test/java/io/lettuce/SslTest.java [hunk 14]
```diff
@@ -217,6 +251,139 @@ public void sslWithVerificationWillFail() {
...
+        SslOptions sslOptions = SslOptions.builder() //
+                .openSslProvider() //
+                .build();
+        setOptions(sslOptions);
+
+        verifyMasterSlaveConnection(MASTER_SLAVE_URIS_VERIFY);
+    }
+
+    @Test
+    public void masterSlavePingBeforeActivate() {
```
Comment (mp911de) [inline src/test/java/io/lettuce/SslTest.java]: 👍

File: src/test/java/io/lettuce/SslTest.java [hunk 15]
```diff
@@ -217,6 +251,139 @@ public void sslWithVerificationWillFail() {
...
+
+        redisClient.setOptions(ClientOptions.builder().pingBeforeActivateConnection(true).build());
+
+        verifyMasterSlaveConnection(MASTER_SLAVE_URIS_NO_VERIFY);
+    }
+
+    @Test
+    public void masterSlaveSslWithReconnect() throws Exception {
+        RedisCommands<String, String> connection =
```
Comment (mp911de) [inline src/test/java/io/lettuce/SslTest.java]: 👍

File: src/test/java/io/lettuce/SslTest.java [hunk 16]
```diff
@@ -217,6 +251,139 @@ public void sslWithVerificationWillFail() {
...
+                MasterSlave.connect(redisClient, StringCodec.UTF8, MASTER_SLAVE_URIS_NO_VERIFY).sync();
+        connection.quit();
+        Thread.sleep(200);
+        assertThat(connection.ping()).isEqualTo("PONG");
+        connection.getStatefulConnection().close();
+    }
+
+    @Test(expected = RedisConnectionException.class)
+    public void masterSlaveSslWithVerificationWillFail() {
```
Comment (mp911de) [inline src/test/java/io/lettuce/SslTest.java]: 👍

File: src/test/java/io/lettuce/SslTest.java [hunk 17]
```diff
@@ -283,6 +450,22 @@ public void pubSubSslAndBreakConnection() throws Exception {
...
 
+
+    private static RedisURI.Builder sslURIBuilder(int portOffset) {
+        return RedisURI.Builder.redis(host(), sslPort(portOffset))
+                .withSsl(true);
+    }
+
+    private static RedisURI.Builder masterSlaveSSLURIBuilder(int portOffset) {
+        return sslURIBuilder(portOffset)
+                // TODO - why do the master/slave tests need this but the standalone ones don't???
```
Comment (mp911de) [inline src/test/java/io/lettuce/SslTest.java]: What happens if you omit this one?

File: src/test/java/io/lettuce/core/cluster/RedisClusterPasswordSecuredSslTest.java (hunk 3 above)
Comment (acmcelwee) [reply src/test/java/io/lettuce/core/cluster/RedisClusterPasswordSecuredSslTest.java]: They aren't. As I dug around I noticed a few intellij inspection warnings that seemed easy to clean up. I'll back them out.

File: src/test/java/io/lettuce/SslTest.java (hunk 17 above)
Comment (acmcelwee) [reply src/test/java/io/lettuce/SslTest.java]: The `SslTest` take a looong time to run (like 400+ seconds vs the 5-10 that they're at after my current changes). There's probably a subtle difference with how the initial connections are setup and how long we wait on connection failures. I haven't figured it out yet, but I do think it should be fixed before we call this ready to merge.

**Comment by acmcelwee** [comment]
I'll clean up the redundant stuff this AM and back out the static builder changes that shouldn't have been in this PR.

File: src/test/java/io/lettuce/SslTest.java (hunk 4 above)
Comment (acmcelwee) [reply src/test/java/io/lettuce/SslTest.java]: My first thought here was that there's enough subtlety that it deserves extra test coverage, but I guess at the end of the day, they all end up just creating one or more `StatefulRedisConnection`s, so I can get onboard w/ them being covered. I'll clean those up now.

File: Makefile (hunk 1 above)
Comment (acmcelwee) [reply Makefile]: ✅ 2138906cc270e733be7a24d2544bd10e3d00d308

File: src/test/java/io/lettuce/SslTest.java (hunk 17 above)
Comment (acmcelwee) [reply src/test/java/io/lettuce/SslTest.java]: This is all that's left of the issues we've identified, so I'll try to figure this out today.

Comment (acmcelwee) [reply src/test/java/io/lettuce/SslTest.java]: Found it. Check out the [changes in StaticMasterSlaveTopologyProvider.java](https://github.com/lettuce-io/lettuce-core/pull/684/files#diff-7081f59597866c51272e8e6ba44922b5) for details.

File: src/main/java/io/lettuce/core/masterslave/StaticMasterSlaveTopologyProvider.java [hunk 18]
```diff
@@ -77,6 +78,13 @@ public StaticMasterSlaveTopologyProvider(RedisClient redisClient, Iterable<Redis
         Flux<RedisURI> uris = Flux.fromIterable(redisURIs);
         Mono<List<RedisNodeDescription>> nodes = uris.flatMap(uri -> getNodeDescription(connections, uri)) //
                 .collectList() //
+                .map((nodeDescriptions) -> {
+                    if (nodeDescriptions.isEmpty()) {
+                        throw new RedisConnectionException(String.format("Failed to connect to any nodes in %s", redisURIs));
```
Comment (acmcelwee) [inline src/main/java/io/lettuce/core/masterslave/StaticMasterSlaveTopologyProvider.java]: On second thought, this functionality could be debated on what the desired behavior is. I've asserted what _I_ think makes the most sense in the [two new tests I added](https://github.com/acmcelwee/lettuce-core/blob/434a385d4e5288b8b4ff68bef849457887a1d535/src/test/java/io/lettuce/SslTest.java#L344-L366).

If you think that it should be a hard fail on a failure to connect to _any_ node in the initial connection setup, we'd need to do something different. LMK what you think, and I'm happy to implement to continue moving this PR along.

Comment (mp911de) [reply src/main/java/io/lettuce/core/masterslave/StaticMasterSlaveTopologyProvider.java]: Thanks for raising this issue. I think you could argue in both ways, hard fail on any failure vs. fail only if no connection could be established at all. Having this safeguard in place is beneficial. Please do not use `.map(…)` but rather `flatMap(…)` and return `Mono.error(…)`. Throwing exceptions in reactive operators isn't a proper way to communicate failures.

However, this question raises another one and it became obvious why there is no SSL support for Master/Slave right now: Redis does not know about SSL (which require a tunnel/proxy), Redis topology discovery reports the actual hosts and ports. For SSL with Sentinel and Master/Slave auto-discovery we need some sort of mapper function to make it work. The reported addresses must be mapped to an URI that points to the actual SSL endpoint.

For now, I'd keep the mapper aside and introduce such a Mapper API in a new ticket.

Comment (acmcelwee) [reply src/main/java/io/lettuce/core/masterslave/StaticMasterSlaveTopologyProvider.java]: `.map(...)` converted to `.flatMap(...)` - 78a51f7e930a74da3144349b5012abb12fbea29b

Yeah, I fully understand why the sentinel or master/slave auto-discovery won't work as-is, and I think it'll take a bit of work to make that mapping work the way we'd need it to. I agree that it should live as a separate, isolated unit of work.

That said, I'd imagine that a lot of people will rely on the TLS support in Elasticache or Azure Cache. For us (using Elasticache), the static topology is mostly fine, but I still think it's not _quite_ perfect w/ master/slave because of how elasticache manages their endpoints. Take a cluster, say `cluster0` with two nodes, `node0` and `node1`. Elasticache basically gives me three dns endpoints along the lines of:
* `node0.cluster0.nzcsol.use1.cache.amazonaws.com` - `A` record to `node0`'s ip in one of our private vpc subnets
* `node1.cluster0.nzcsol.use1.cache.amazonaws.com` - `A` record to `node``'s ip in one of our private vpc subnets
* `master.cluster0.nzcsol.use1.cache.amazonaws.com` - `CNAME` record to either `node0` or `node1`, depending on which node is currently the primary node

Now, the problem comes in the fact that all nodes need to be specified at the time of creation of a static master/slave topology, and that the node's role (master or slave) is locked when the client is initially created. I have to specify the the `master.cluster0...` node, otherwise the client won't work in the event of a primary failover. To fully utilize all of the read replicas in the cluster, I also have to specify both of the `node0...` and `node1...` hostnames when specifying the endpoints for the static topology master/slave. That leaves me w/ a list of 3 `knownNodes` in `MasterSlaveConnectionProvider`, because whichever node is currently the master at client creation time ends up with two connections to that it, both marked as `MASTER`. As part of the dns record to reported address work, it might also be good to think about how to dedupe those connections. If we could make the static topology _slightly_ dynamic, that would be awesome. I could imagine a nice middle-ground between, dynamically discovered based on a seed node, and static_ish_ by requiring the full set of nodes listed at creation time but that will re-determine the master node at the time of failover.

**Comment by mp911de** [comment]
Thanks a lot for your pull request and all your work related to this issue. That's merged, polished and ported to all affected branches now. Snapshot builds are available now.

**Comment by acmcelwee** [comment]
Excellent. I'm happy to help. I'll build and push to our company artifactory for now, but what are your plans for the 4.4.x and 5.x branch releases?

**Comment by mp911de** [comment]
I added dates to the bugfix releases for end of February, see https://github.com/lettuce-io/lettuce-core/milestones.
